test/sslapitest.c

```diff
@@ -9229,6 +9229,99 @@ static int test_pluggable_group(int idx)
+
+    if (!TEST_ptr(evpctx) ||
+        !TEST_true(EVP_PKEY_keygen_init(evpctx)) ||
+        !TEST_true(EVP_PKEY_generate(evpctx, &pkey)) ||
+        !TEST_ptr(pkey) ||
+        !TEST_ptr(x509) ||
+        !TEST_true(ASN1_INTEGER_set(X509_get_serialNumber(x509), 1)) ||
+        !TEST_true(X509_gmtime_adj(X509_getm_notBefore(x509), 0)) ||
+        !TEST_true(X509_gmtime_adj(X509_getm_notAfter(x509), 31536000L)) ||
+        !TEST_true(X509_set_pubkey(x509, pkey)) ||
+        !TEST_ptr(name = X509_get_subject_name(x509)) ||
+        !TEST_true(X509_NAME_add_entry_by_txt(name, "C",  MBSTRING_ASC,
+                           (unsigned char *)"CH", -1, -1, 0)) ||
+        !TEST_true(X509_NAME_add_entry_by_txt(name, "O",  MBSTRING_ASC,
+                           (unsigned char *)"test.org", -1, -1, 0)) ||
+        !TEST_true(X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC,
+                           (unsigned char *)"localhost", -1, -1, 0)) ||
+        !TEST_true(X509_set_issuer_name(x509, name)) ||
+        !TEST_true(X509_sign(x509, pkey, EVP_sha1())) ||
+        !TEST_ptr(keybio = BIO_new_file(privkeyfilename, "wb")) ||
+        !TEST_true(PEM_write_bio_PrivateKey(keybio, pkey, NULL, NULL, 0, NULL, NULL)) ||
+        !TEST_ptr(certbio = BIO_new_file(certfilename, "wb")) ||
+        !TEST_true(PEM_write_bio_X509(certbio, x509)))
+        ret = 0;
+
+    EVP_PKEY_free(pkey);
+    X509_free(x509);
+    EVP_PKEY_CTX_free(evpctx);
+    BIO_free(keybio);
+    BIO_free(certbio);
+    return ret;
+}
+
+static int test_pluggable_signature(int idx)
+{
+    SSL_CTX *cctx = NULL, *sctx = NULL;
+    SSL *clientssl = NULL, *serverssl = NULL;
+    int testresult = 0;
+    OSSL_PROVIDER *tlsprov = OSSL_PROVIDER_load(libctx, "tls-provider");
+    OSSL_PROVIDER *defaultprov = OSSL_PROVIDER_load(libctx, "default");
+    char *certfilename = "tls-prov-cert.pem";
+    char *privkeyfilename = "tls-prov-key.pem";
+
+    if (!TEST_ptr(tlsprov) ||
+        !TEST_true(create_cert_key(idx, certfilename, privkeyfilename)))
+        goto end;
+
+    if (!TEST_true(create_ssl_ctx_pair(libctx, TLS_server_method(),
+                                       TLS_client_method(),
+                                       TLS1_3_VERSION,
+                                       TLS1_3_VERSION,
+                                       &sctx, &cctx, certfilename, privkeyfilename))
+            || !TEST_true(create_ssl_objects(sctx, cctx, &serverssl, &clientssl,
+                                             NULL, NULL)))
+        goto end;
+
+    /* This is necessary to pass minimal setup w/o other groups configured */
+    if (!TEST_true(SSL_set1_groups_list(serverssl, "xorgroup"))
+            || !TEST_true(SSL_set1_groups_list(clientssl, "xorgroup")))
+        goto end;
+
+    /* If this connection gets established, it must have been completed
+     * via the tls-provider-implemented "hmacsig" algorithm, testing
+     * both sign and verify functions during handshake.
+     */
```

Comment:
...

Reply:
changed in https://github.com/openssl/openssl/pull/19312/commits/0330203fc25c8b2d08c052ef8070ab6f25d17209